core/controller/src/main/scala/whisk/core/controller/Meta.scala

```diff
@@ -190,77 +448,108 @@ trait WhiskMetaApi extends Directives with PostActionActivation {
-                    // if action name is defined as a string, accept it, else fail request
-                    case Some(JsString(actionName)) =>
-                        info(this, s"'${pkg.name}' maps '${method.name}' to action '${actionName}'")
-                        Future.successful(EntityName(actionName), pkg.parameters)
-                    case _ =>
-                        info(this, s"'${pkg.name}' is missing 'meta' annotation or action name for '${method.name.toLowerCase}'")
-                        Future.failed(RejectRequest(MethodNotAllowed))
-                }
-            } else {
-                warn(this, s"'${pkg.fullyQualifiedName(true)}' is a binding")
-                Future.failed(RejectRequest(MethodNotAllowed))
+            // does package have annotation: meta == true and a
+            // mapping from HTTP verb to action name?
+            val isMetaPackage = pkg.annotations.asBool("meta")
+
+            isMetaPackage flatMap {
+                _ => pkg.annotations.asString(method.name.toLowerCase)
+            } match {
+                // if action name is defined as a string, accept it, else fail request
+                case Some(actionName) =>
+                    info(this, s"'${pkg.name}' maps '${method.name}' to action '${actionName}'")
+                    Future.successful(EntityName(actionName), pkg.parameters)
+
+                case _ =>
+                    info(this, s"'${pkg.name}' is missing 'meta' annotation or action name for '${method.name.toLowerCase}'")
+                    isMetaPackage match {
+                        case Some(true) => Future.failed(RejectRequest(MethodNotAllowed))
+                        case _          => Future.failed(RejectRequest(NotFound))
+                    }
             }
         }
     }
 
-    protected def actionLookup(pkgName: FullyQualifiedEntityName, actionName: EntityName)(
+    /**
+     * Checks if an action is exported (i.e., carries the required annotation).
+     */
+    private def confirmExportedAction(actionLookup: Future[WhiskAction])(
         implicit transid: TransactionId): Future[WhiskAction] = {
-        val docid = pkgName.add(actionName).toDocId
-        WhiskAction.get(entityStore, docid) recoverWith {
-            case _: ArtifactStoreException | DeserializationException(_, _, _) =>
-                // the action doesn't exist or is corrupted but the package stated otherwise
-                // so treat this as an internal error
-                Future.failed(RejectRequest(InternalServerError))
+        actionLookup flatMap { action =>
+            action.annotations.asBool("web-export").filter(b => b).map {
+                _ =>
+                    info(this, s"${action.fullyQualifiedName(true)} is exported")
+                    Future.successful(action)
+            } getOrElse {
+                info(this, s"${action.fullyQualifiedName(true)} not exported")
+                Future.failed(RejectRequest(NotFound))
+            }
```

Comment:
Minor scalaism nit: You can convert an `Option` to a `Boolean` by using `exists`:

```
if(action.annotations.asBool("web-export").exists(identity)) {
    info(this, s"${action.fullyQualifiedName(true)} is exported")
    Future.successful(action)
} else {
    info(this, s"${action.fullyQualifiedName(true)} not exported")
    Future.failed(RejectRequest(NotFound))
}
```

You decide what you like more.